File: rb/test/page_entity_test.rb
# Page entity test

require "minitest/autorun"
require "json"
require_relative "../Statuspage_sdk"
require_relative "runner"

class PageEntityTest < Minitest::Test
  def test_create_instance
    testsdk = StatuspageSDK.test(nil, nil)
    ent = testsdk.Page(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = page_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list", "update", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "page." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set STATUSPAGE_TEST_PAGE_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    page_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.page")))
    page_ref01_data = nil
    if page_ref01_data_raw.length > 0
      page_ref01_data = Helpers.to_map(page_ref01_data_raw[0][1])
    end

    # LIST
    page_ref01_ent = client.Page(nil)
    page_ref01_match = {}

    page_ref01_list_result = page_ref01_ent.list(page_ref01_match, nil)
    assert page_ref01_list_result.is_a?(Array)

    # UPDATE
    page_ref01_data_up0_up = {
      "id" => page_ref01_data["id"],
    }

    page_ref01_markdef_up0_name = "branding"
    page_ref01_markdef_up0_value = "Mark01-page_ref01_#{setup[:now]}"
    page_ref01_data_up0_up[page_ref01_markdef_up0_name] = page_ref01_markdef_up0_value

    page_ref01_resdata_up0_result = page_ref01_ent.update(page_ref01_data_up0_up, nil)
    page_ref01_resdata_up0 = Helpers.to_map(page_ref01_resdata_up0_result)
    assert !page_ref01_resdata_up0.nil?
    assert_equal page_ref01_resdata_up0["id"], page_ref01_data_up0_up["id"]
    assert_equal page_ref01_resdata_up0[page_ref01_markdef_up0_name], page_ref01_markdef_up0_value

    # LOAD
    page_ref01_match_dt0 = {
      "id" => page_ref01_data["id"],
    }
    page_ref01_data_dt0_loaded = page_ref01_ent.load(page_ref01_match_dt0, nil)
    page_ref01_data_dt0_load_result = Helpers.to_map(page_ref01_data_dt0_loaded)
    assert !page_ref01_data_dt0_load_result.nil?
    assert_equal page_ref01_data_dt0_load_result["id"], page_ref01_data["id"]

  end
end

def page_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "page", "PageTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = StatuspageSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["page01", "page02", "page03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["STATUSPAGE_TEST_PAGE_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "STATUSPAGE_TEST_PAGE_ENTID" => idmap,
    "STATUSPAGE_TEST_LIVE" => "FALSE",
    "STATUSPAGE_TEST_EXPLAIN" => "FALSE",
    "STATUSPAGE_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["STATUSPAGE_TEST_PAGE_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["STATUSPAGE_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["STATUSPAGE_APIKEY"],
      },
      extra || {},
    ])
    client = StatuspageSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["STATUSPAGE_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["STATUSPAGE_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/page_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/statuspage-sdk/go"
	"github.com/voxgig-sdk/statuspage-sdk/go/core"

	vs "github.com/voxgig-sdk/statuspage-sdk/go/utility/struct"
)

func TestPageEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Page(nil)
		if ent == nil {
			t.Fatal("expected non-nil PageEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"page": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Page(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Page(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := pageBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "page." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set STATUSPAGE_TEST_PAGE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		pageRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.page", setup.data)))
		var pageRef01Data map[string]any
		if len(pageRef01DataRaw) > 0 {
			pageRef01Data = core.ToMapAny(pageRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = pageRef01Data

		// LIST
		pageRef01Ent := client.Page(nil)
		pageRef01Match := map[string]any{}

		pageRef01ListResult, err := pageRef01Ent.List(pageRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, pageRef01ListOk := pageRef01ListResult.([]any)
		if !pageRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", pageRef01ListResult)
		}

		// UPDATE
		pageRef01DataUp0Up := map[string]any{
			"id": pageRef01Data["id"],
		}

		pageRef01MarkdefUp0Name := "branding"
		pageRef01MarkdefUp0Value := fmt.Sprintf("Mark01-page_ref01_%d", setup.now)
		pageRef01DataUp0Up[pageRef01MarkdefUp0Name] = pageRef01MarkdefUp0Value

		pageRef01ResdataUp0Result, err := pageRef01Ent.Update(pageRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		pageRef01ResdataUp0 := core.ToMapAny(pageRef01ResdataUp0Result)
		if pageRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if pageRef01ResdataUp0["id"] != pageRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if pageRef01ResdataUp0[pageRef01MarkdefUp0Name] != pageRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", pageRef01MarkdefUp0Name, pageRef01ResdataUp0[pageRef01MarkdefUp0Name])
		}

		// LOAD
		pageRef01MatchDt0 := map[string]any{
			"id": pageRef01Data["id"],
		}
		pageRef01DataDt0Loaded, err := pageRef01Ent.Load(pageRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		pageRef01DataDt0LoadResult := core.ToMapAny(pageRef01DataDt0Loaded)
		if pageRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if pageRef01DataDt0LoadResult["id"] != pageRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func pageBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "page", "PageTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read page test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse page test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"page01", "page02", "page03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("STATUSPAGE_TEST_PAGE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"STATUSPAGE_TEST_PAGE_ENTID": idmap,
		"STATUSPAGE_TEST_LIVE":      "FALSE",
		"STATUSPAGE_TEST_EXPLAIN":   "FALSE",
		"STATUSPAGE_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["STATUSPAGE_TEST_PAGE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["STATUSPAGE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["STATUSPAGE_APIKEY"],
			},
			extra,
		})
		client = sdk.NewStatuspageSDK(core.ToMapAny(mergedOpts))
	}

	live := env["STATUSPAGE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["STATUSPAGE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
